aten/src/ATen/native/quantized/cpu/qnnpack_fc.cpp

```diff
@@ -0,0 +1,98 @@
+#include "init_qnnpack.h"
+
+#include <ATen/ATen.h>
+#include <ATen/core/Type.h>
+#include <ATen/core/op_registration/op_registration.h>
+#include <ATen/quantized/Quantizer.h>
+#include <ATen/Config.h>
+
+namespace at { namespace native {
+namespace {
+
+class QNNPACKFullyConnected final : public c10::OperatorKernel {
+ public:
+  Tensor operator()(at::Tensor input,
+      at::Tensor weight,
+      at::Tensor bias,
+      double output_scale,
+      int64_t output_zero_point) {
+
+    TORCH_CHECK(input.dim() >= 2, "Input tensor rank should be >= 2");
+
+    Tensor input_contig = input.contiguous();
+
+    // C(output) = A(input_contig) x B(weight), where C, A, B are M x N, M x K, N x K
+    // matrices, respectively.
+    int64_t rows_a = input_contig.size(0);
+    int64_t cols_a = 1;
+    for (size_t i = 1; i < input_contig.dim(); ++i) {
+      cols_a *= input_contig.size(i);
+    }
+
+    int64_t rows_b = weight.size(0);
+
+    TORCH_CHECK(cols_a == weight.size(1),
+        "qnnpack_fc(): input size does not match weight dimension 1 size: got ",
+        cols_a, " but expected ", weight.size(1));
+
+    TORCH_CHECK(!bias.defined() || (bias.ndimension() == 1 && bias.size(0) == rows_b),
+        "qnnpack_fc(): Given weight of size ", weight.sizes(),
+        ", expected bias to be 1-dimensional with ", rows_b, " elements",
+        ", but got bias of size ", bias.sizes(), " instead");
+
+    initQNNPACK();
+    qnnp_operator_t qnnpackOperator_{nullptr};
+
+    // Allocate output Tensor and a buffer for QNNPACK to use
+    Tensor output = at::_empty_affine_quantized(
+      {rows_a, rows_b},
+      at::device(kCPU).dtype(kQUInt8),
+      output_scale,
+      output_zero_point);
+
+    // QNNPACK expects both weights and inputs to be uint8
+    const qnnp_status createStatus = qnnp_create_fully_connected_nc_q8(
```

Comment:
how expensive is creating qnnpack op? we should consider creating a cache of those if needed

Reply:
It's expensive, and it's also conceptually a weight-packing step - cf. https://github.com/pytorch/QNNPACK/blob/b7bacb1899e6fa3a934c1dd6128096f2e1abf071/src/fully-connected.c#L94

Reply:
Conceptually I think this approach is wrong - like, QNNPACK is designed in a C2/TF/MXNet world of stateful operators. Essentially, we call `qnnp_create_*` in the constructor, which does packing/allocation/one-time setup. These `create_` instances can't be 'cached' because they contain state about the weights they were instantiated with. You then call `setup + run` at the `Op::Run` stage. 

Conceptually it's not really reasonable (from the perspective of achieving optimal perf) to stuff it all into the `run` call - it'll work fine but it's not right. We should either refactor QNNPACK to be more functionally oriented (which would be fairly mechanical but a decent chunk of work), or rethink this approach to match the existing calling pattern for QNNPACK + C2.

Reply:
I recollect this being an initial step. We should have two ops eventually:

 qnnpack_state =_qnnpack_linear_create_state(weight and other args). (This calls create)

and then have the actual op: qnnpack_linear() that performs the setup and run.

Reply:
Yes, that is the plan. The plan was to submit this PR as an initial step and then work on perf optimizations by implementing caching of the create step. @dzhulgakov is that okay?

Reply:
Yeah, Andrew is right. Unfortunately, qnnpack itself is too static and it's easy to "break it" even in c2 bindings. For example, it bakes the quantization params for the input in op creation here: https://github.com/pytorch/pytorch/blob/28b5df1c8f8133f6bfd6326d6c83ef2578ab4518/caffe2/operators/quantized/int8_fc_op.h#L57 - so if someone calls the op second time with differently quantized input it'd break (which theoretically possible even in c2 world as we pass quantization parameters not statically but through input tensors, though it rarely happens in practice). The reason for it is that it pre-scales the weights accordingly (which is faster I suppose).

so even if we do _qnnpack_linear_create_state we'd need to pass quantization parameters for the input to it statically. It's doable as a JIT transform in most cases. I guess the best we can do as an isolated transformation is to also pass "expected" input quantization params to _qnnpack_linear_create_state and later error out if the params don't match the ones on the input tensor passed to qnnpack_linear

another problem with qnnpack is that it's not really reentrant - _setup method just modifies the "packed" structure in-place and thus the same state can't be accessed from different threads. That's pretty easy to fix though - pretty much we should change these params to be function arguments instead of fields in the object: https://github.com/pytorch/QNNPACK/blob/e95863af039a8c48e0b820a154b02f7e3d7f0690/src/fully-connected.c#L131